File: Individual Assignment 1 - Matrix Multiplication Benchmark/matrix_naive.py
import sys
import time
import random
import os
import psutil

def init_matrix_zero(n):
    return [[0.0 for _ in range(n)] for _ in range(n)]

def set_matrix_zero(M):
    n = len(M)
    for i in range(n):
        for j in range(n):
            M[i][j] = 0.0

def rand_matrix(M):
    n = len(M)
    for i in range(n):
        for j in range(n):
            M[i][j] = random.random()

def mult_naive(a, b, c):
    n = len(a)
    for i in range(n):
        for j in range(n):
            for k in range(n):
                c[i][j] += a[i][k] * b[k][j]

def main():
    if len(sys.argv) < 3:
        print("Usage: python matrix_naive.py <n> <runs>")
        sys.exit(1)

    n = int(sys.argv[1])
    runs = int(sys.argv[2])

    process = psutil.Process(os.getpid())
    memory_start = process.memory_info().rss

    a = init_matrix_zero(n)
    b = init_matrix_zero(n)
    c = init_matrix_zero(n)

    memory_stop = process.memory_info().rss

    time_start = time.time()

    for _ in range(runs):
        rand_matrix(a)
        rand_matrix(b)
        set_matrix_zero(c)
        mult_naive(a, b, c)

    time_stop = time.time()

    memory_diff = (memory_stop - memory_start) / (1024 * 1024)
    time_diff = time_stop - time_start

    print(f"Python: n={n}, runs={runs}, time={time_diff:.3f}s, "
          f"timePerRun={time_diff/runs:.3f}s, memoryUsed={memory_diff:.2f}MB")

if __name__ == "__main__":
    main()

File: Individual Assignment 1 - Matrix Multiplication Benchmark/MatrixNaive.java
import java.util.Random;

public class MatrixNaive {
  static double[][] initMatrixZero(int n) {
    double[][] M = new double[n][n];
    setMatrixZero(M);
    return M;
  }

  static void setMatrixZero(double[][] M) {
    int n = M.length;
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < n; j++) {
        M[i][j] = 0;
      }
    }
  }

   static void randMatrix(double[][] M) {
    Random random = new Random();
    int n = M.length;
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < n; j++) {
        M[i][j] = random.nextDouble();
      }
    }
  }

 static void multNaive(double[][] a, double[][] b, double[][] c){
    int n = a.length;
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < n; j++) {
        for (int k = 0; k < n; k++) {
          c[i][j] += a[i][k] * b[k][j];
        }
      }
    }
  }

	public static void main(String[] args) {
    if (args.length < 2) {
      System.err.println("Usage: java matrixNaive <n> <runs>");
      System.exit(1);
    }

    int n = Integer.parseInt(args[0]);
    int runs = Integer.parseInt(args[1]);

    Runtime runtime = Runtime.getRuntime();
    double memoryStart = runtime.totalMemory() - runtime.freeMemory();

    double[][] a = initMatrixZero(n);
    double[][] b = initMatrixZero(n);
    double[][] c = initMatrixZero(n);

    double memoryStop = runtime.totalMemory() - runtime.freeMemory();

		long timeStart = System.currentTimeMillis();

    for(int i = 0; i < runs; i++){
      randMatrix(a);
      randMatrix(b);
      setMatrixZero(c);
      multNaive(a, b, c);
    }

		long timeStop = System.currentTimeMillis();

    double memoryDiff = (memoryStop - memoryStart) / (1024 * 1024);
    double timeDiff = ((timeStop - timeStart) * 1e-3);

    System.out.printf("Java: n=%d, runs=%d, time=%.3fs, timePerRun=%.3fs, memoryUsed=%.2fMB\n", n, runs, timeDiff, timeDiff/runs, memoryDiff);
	}
}

